pkg/auth/providers/publicapi/login.go

```diff
@@ -208,9 +213,61 @@ func (h *loginHandler) createLoginToken(request *types.APIContext) (v3.Token, st
 		if err != nil {
 			return v3.Token{}, "", "", err
 		}
+
+		if err := h.createClusterAuthTokenIfNeeded(token, tokenValue); err != nil {
+			logrus.Infof("Failed to create cluster auth token for cluster [%s], cluster auth endpoint may fail: %v", token.ClusterName, err)
+		}
 		return *token, tokenValue, responseType, nil
 	}
 
 	rToken, unhashedTokenKey, err := h.tokenMGR.NewLoginToken(currUser.Name, userPrincipal, groupPrincipals, providerToken, ttl, description)
 	return rToken, unhashedTokenKey, responseType, err
 }
+
+// createClusterAuthTokenIfNeeded checks if local cluster auth endpoint is enabled. If it is, a cluster auth token
+// is enabled.
+func (h *loginHandler) createClusterAuthTokenIfNeeded(token *v3.Token, tokenValue string) error {
+	clusterID := token.ClusterName
+	cluster, err := h.clusterLister.Get("", clusterID)
+	if err != nil {
+		return err
+	}
+
+	if !cluster.Spec.LocalClusterAuthEndpoint.Enabled {
+		return nil
+	}
+	clusterConfig, err := clustermanager.ToRESTConfig(cluster, h.scaledContext)
+	if err != nil {
+		return err
+	}
+
+	clusterContext, err := config.NewUserContext(h.scaledContext, *clusterConfig, clusterID)
+	if err != nil {
+		return err
+	}
+
+	clusterAuthToken, err := common.NewClusterAuthToken(token, tokenValue)
+	if err != nil {
+		return err
+	}
+
+	_, err = clusterContext.Cluster.ClusterAuthTokens("").Controller().Lister().Get("cattle-system", clusterAuthToken.Name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	if err == nil {
+		err = clusterContext.Cluster.ClusterAuthTokens("cattle-system").Delete(clusterAuthToken.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	if _, err = clusterContext.Cluster.ClusterAuthTokens("cattle-system").Create(clusterAuthToken); err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
I'm confused why this broken now with token hashing. I get that on the kubeconfig generation side, we added it in actions_kubeconfig.go because it has to be done at token generation time before it gets hashed. And we are roughly doing the same thing here, but we weren't creating an auth token here before, so was it already broken ?

Reply:
Before a handler (2.5) was taking care of it, after I made changes the handles can't do it anymore and I forgot to add it for login token.